jerry-core/context/context.h

```diff
@@ -0,0 +1,102 @@
+/* Copyright 2016 Samsung Electronics Co., Ltd.
+ * Copyright 2016 University of Szeged.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Memory context for JerryScript
+ */
+#ifndef CONTEXT_H
+#define CONTEXT_H
+
+#include "jrt.h"
+#include "mem-allocator.h"
+#include "mem-config.h"
+
+/** \addtogroup context Jerry context
+ * @{
+ *
+ * \addtogroup context Context
+ * @{
+ */
+
+/**
+ * Calculate heap area size, leaving space for a pointer to the free list
+ */
+#define MEM_HEAP_AREA_SIZE (MEM_HEAP_SIZE - MEM_ALIGNMENT)
+
+/**
+ * Heap structure
+ */
+typedef struct
+{
+  mem_heap_free_t first; /**< first node in free region list */
+  uint8_t area[MEM_HEAP_AREA_SIZE]; /**< heap area */
+} mem_heap_t;
+
+/**
+ * JerryScript context
+ *
+ * The purpose of this header is storing
+ * all global variables for Jerry
+ */
+typedef struct
+{
+  /**
+   * Memory manager part.
+   */
+  size_t mem_heap_allocated_size; /**< size of allocated regions */
+  size_t mem_heap_limit; /**< current limit of heap usage, that is upon being reached,
+                          *   causes call of "try give memory back" callbacks */
+  mem_heap_free_t *mem_heap_list_skip_p; /**< This is used to speed up deallocation. */
+  mem_pools_chunk_t *mem_free_chunk_p; /**< list of free pool chunks */
+  mem_try_give_memory_back_callback_t mem_try_give_memory_back_callback; /**< The 'try to give memory back' callback */
+
+#ifdef MEM_STATS
+  mem_heap_stats_t mem_heap_stats; /**< heap's memory usage statistics */
+  mem_pools_stats_t mem_pools_stats; /**< pools' memory usage statistics */
+#endif /* MEM_STATS */
+
+#ifdef JERRY_VALGRIND_FREYA
+  bool valgrind_freya_mempool_request; /**< Tells whether a pool manager
+                                        *   allocator request is in progress */
+#endif /* JERRY_VALGRIND_FREYA */
+} jerry_context_t;
+
+/**
+ * Jerry global context.
+ */
+extern jerry_context_t jerry_global_context;
+
+/**
+ * Jerry global heap.
+ */
+extern mem_heap_t jerry_global_heap;
```

Comment:
>2) The mem_heap_t is very much needed, since the two fields are tightly coupled. Actually the "first" is the first 8 bytes of the heap, and provides nicer access to this 8 bytes than casting uint8_t* all the time.

I don't think that this is right/correct. I've tried to trace down all uses of the `first` field and it is never used as part of the heap. I cannot find a place where it would be referred to via a compact pointer. It is never used in a more tightly coupled way than `mem_heap_list_skip_p`, which is in the "regular" context object.

>3) What do you expect here?

My original suggestion (the first in this thread) was to have a pointer in the context object to a heap area, which is "somewhere"; instead of having a globally allocated byte array for the heap. You've argued against it because of performance reasons. That's why I asked for benchmark results.